addons/stock/report/stock_forecasted.py

```diff
@@ -269,11 +268,37 @@ def _reconcile_out_with_ins(lines, out, ins, demand, product_rounding, only_matc
         )
 
         outs = self.env['stock.move'].search(out_domain, order='reservation_date, priority desc, date, id')
+        ins = self.env['stock.move'].search(in_domain, order='priority desc, date, id')
+        # Prewarm cache with rollups
+        outs._rollup_move_origs_fetch()
+        ins._rollup_move_dests_fetch()
+
+        linked_moves_per_out = {}
+        ins_ids = set(ins._ids)
+        for out in outs:
+            linked_move_ids = out._rollup_move_origs() - ins_ids
+            linked_moves_per_out[out] = self.env['stock.move'].browse(linked_move_ids)
+        
+        # Gather all relevant moves, ins + outs + linked moves
+        all_linked_move_ids = {_id for _ids in linked_moves_per_out.values() for _id in _ids._ids}
+        all_linked_moves = self.env['stock.move'].browse(all_linked_move_ids)
+        all_moves = (outs | ins | all_linked_moves)
+
+        # Prewarm cache with sibling move's state/quantity
+        all_linked_moves.fetch(['move_orig_ids'])
+        all_linked_moves.move_orig_ids.fetch(['move_dest_ids'])
+        all_linked_moves.move_orig_ids.move_dest_ids.fetch(['state', 'quantity'])
+
+        # Share prefetch ids among all relevant moves for performance
+        outs = outs.with_prefetch(all_moves._prefetch_ids)
+        ins = ins.with_prefetch(all_moves._prefetch_ids)
+        for out, linked_moves in linked_moves_per_out.items():
+            linked_moves_per_out[out] = linked_moves.with_prefetch(all_moves._prefetch_ids)
+
```

Comment:
```suggestion
        # Prewarm cache with sibling move's state/quantity
        all_linked_moves.fetch(['move_orig_ids'])
        all_linked_moves.move_orig_ids.fetch(['move_dest_ids'])
        all_linked_moves.move_orig_ids.move_dest_ids.fetch(['state', 'quantity'])

        # Share prefetch ids among all relevant moves for performance
        for out, linked_moves in linked_moves_per_out.items():
            linked_moves_per_out[out] = linked_moves.with_prefetch(all_linked_moves._prefetch_ids)
```
It seems a bit overkill to put everything in the `_prefetch_ids` of `outs`, `ins` and `linked_moves_per_out`. Sure it reduces the number of queries but as a tradeoff you may run into an issue with `expand_ids` or `_in_cache_without`. Also, `BaseModel.__iter__` has an optimization only when `_ids is _prefetch_ids`.

What do you think about leaving `ins` and `outs` as returned by the calls to `search` and only setting the `_prefetch_ids` of `linked_moves_per_out` to `all_linked_moves` ? I may be wrong but it seems that setting the `_prefetch_ids` is mostly useful for the call to `_get_out_move_reserved_data`.

Reply:
Good point. I was thinking since they all fetch similar fields and were from the same table it may be OK to lump them all together. But yeah that might not be the best idea for other reasons as you explain. I will test with just setting the _prefetch_ids of linked_moves_per_out and see how that does.

Reply:
Implemented in amend.